src/boot/efi/linux.c

```diff
@@ -13,83 +13,22 @@
 
 #include "initrd.h"
 #include "linux.h"
+#include "missing_efi.h"
 #include "pe.h"
 #include "util.h"
 
-static EFI_LOADED_IMAGE_PROTOCOL *loaded_image_free(EFI_LOADED_IMAGE_PROTOCOL *img) {
-        if (!img)
-                return NULL;
-        mfree(img->LoadOptions);
-        return mfree(img);
-}
-
-static EFI_STATUS loaded_image_register(
-                const char *cmdline, UINTN cmdline_len,
-                const void *linux_buffer, UINTN linux_length,
-                EFI_HANDLE *ret_image) {
-
-        EFI_LOADED_IMAGE_PROTOCOL *loaded_image = NULL;
-        EFI_STATUS err;
-
-        assert(cmdline || cmdline_len > 0);
-        assert(linux_buffer && linux_length > 0);
-        assert(ret_image);
-
-        /* create and install new LoadedImage Protocol */
-        loaded_image = xnew(EFI_LOADED_IMAGE_PROTOCOL, 1);
-        *loaded_image = (EFI_LOADED_IMAGE_PROTOCOL) {
-                .ImageBase = (void *) linux_buffer,
-                .ImageSize = linux_length
-        };
-
-        /* if a cmdline is set convert it to UCS2 */
-        if (cmdline) {
-                loaded_image->LoadOptions = xstra_to_str(cmdline);
-                loaded_image->LoadOptionsSize = strsize16(loaded_image->LoadOptions);
-        }
-
-        /* install a new LoadedImage protocol. ret_handle is a new image handle */
-        err = BS->InstallMultipleProtocolInterfaces(
-                        ret_image,
-                        &LoadedImageProtocol, loaded_image,
-                        NULL);
-        if (err != EFI_SUCCESS)
-                loaded_image = loaded_image_free(loaded_image);
-
-        return err;
-}
-
-static EFI_STATUS loaded_image_unregister(EFI_HANDLE loaded_image_handle) {
-        EFI_LOADED_IMAGE_PROTOCOL *loaded_image;
+EFI_STATUS load_trusted_image(EFI_HANDLE parent, const void *source, size_t len, EFI_HANDLE *ret_image) {
         EFI_STATUS err;
 
-        if (!loaded_image_handle)
-                return EFI_SUCCESS;
-
-        /* get the LoadedImage protocol that we allocated earlier */
-        err = BS->OpenProtocol(
-                        loaded_image_handle, &LoadedImageProtocol, (void **) &loaded_image,
-                        NULL, NULL, EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+        /* We want to be nice and provide a device path. Since the kernel is embedded into our stub we can
+         * just reuse our own device path. */
+        EFI_DEVICE_PATH *loaded_dp;
+        err = BS->HandleProtocol(
+                        parent, &(EFI_GUID) EFI_LOADED_IMAGE_DEVICE_PATH_PROTOCOL_GUID, (void **) &loaded_dp);
```

Comment:
I guess we could also use the MemoryMapped device path but I'm not sure if it would actually make a difference.

Reply:
Afaik the device path we pass really doesn't matter. We could pass NULL even and the firmware should end up doing the right thing (and the zboot should still work then).

Also, I don't think a MemoryMapped device path node is really the right match…

Reply:
Yeah, passing the existing one should be fine, I don't have a preference here

Reply:
It is LoadImage()'s job to install the EFI_LOADED_IMAGE_DEVICE_PATH_PROTOCOL so you shouldn't need to do anything here. (The firmware will just install a NULL protocol if no device path was provided)

Reply:
I'm aware you can just pass NULL to LoadImage, as I've started out with that. But I also think it's nicer to pass something, especially since we can pass something meaningful. It also helps/is needed for the security hook to identify the image we're loading.

Reply:
True, but now there are two different images in the protocol database with the same EFI_LOADED_IMAGE_DEVICE_PATH_PROTOCOL, which I don't think is the correct approach. 

If you don't want to construct a device path that properly describes the encapsulated PE/COFF image, it is better to pass nothing at all rather than a device path that you know describes something different.